[Backtracing] Implement API per SE-0419

al45tair opened the pull request:
Implement the API described in SE-0419, in a new module `Runtime` that you can import from your Swift program.

Additionally, made symbolicated backtraces approximately 5x faster on Linux.

rdar://124913332

al45tair committed:
[Backtracing] Remove support for implicit import of _Backtracing.

We're going to rename the module to Runtime, and it isn't going to be an
implicit import, so we don't need any of this.

rdar://124913332

al45tair committed:
[Backtracing] Rename _Backtracing to Runtime.

Move the backtracing code into a new Runtime module.  This means renaming
the Swift Runtime's CMake target because otherwise there will be a name
clash.

rdar://124913332

al45tair committed:
[Backtracing] Also try `CFStringGetCharactersPtr` when converting `CFString`s.

We already try using `CFStringGetCStringPtr` to see if we can get ASCII;
we should also try `CFStringGetCharactersPtr` as the native encoding of
`CFNSString` is very often UTF-16.  While we're at it, try asking explicitly
for UTF-8 as well.  That won't work for `CFNSString` because AFAIK it's
currently never encoded that way, but it might work if something hands us
a Swift-backed string or if some future implementation of `CFString`
happens to support UTF-8 encoded data.

rdar://124913332

al45tair committed:
[Backtracing] Add ImageMap instead of just using an Array.

We want to be able to efficiently serialise lists of images, and to do so
it makes most sense to create a separate `ImageMap` type.  This also provides
a useful place to put methods to e.g. find an image by address or by build
ID.

rdar://124913332

al45tair committed:
[Backtracing] Implement Codable for Backtrace.

Add support for Codable to Bactkrace.

rdar://124913332

al45tair committed:
[Backtracing] Add support for repeats to Compact Backtrace Format.

This lets us repeat a frame over and over; it only works for a single
frame, but that's the common case for infinite recursions.

rdar://124913332

al45tair commented:
The first changeset, "Update module maps", is [the subject of a separate PR](https://github.com/swiftlang/swift/pull/78514).

al45tair commented:
Hmmm. Looks like maybe my change to add `swift-backtrace` as an explicit dependency in `test/CMakeLists.txt` was incorrect.

al45tair committed:
[Backtracing][Tests] Fix the name of the swift-backtrace dependency.

Because it's a target executable, the name of the dependency is suffixed
with the SDK.

rdar://124913332

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing] Tweak tests slightly.

`JSONEncoder` by default will escape slashes, which results in a string
that isn't technically valid Base64.  Since that behaviour is optional,
turn it off, and at the same time tell it to output in lexical key
order, which makes the test slightly simpler (no `CHECK-DAG` needed).

Also fixed a typo in `test_swift.py`

rdar://124913332

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing] Fix dependency for static swift-backtrace.

We'd missed renaming a CMake variable, which led to link errors.

rdar://124913332

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing] CMake fixes.

Remove a `-disable-implicit-backtracing-module-import` that was added
in `Runtimes/Core/CMakeLists.txt`, and also remove the reference to
`swiftCxxStdlib` from `RuntimeModule` as that causes the build to fail
when attempting to build the static version of `swift-backtrace` on
Linux.

rdar://124913332

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing] Fix Python test failure.

We needed to add `--enable-runtime-module` in the Python test scripts.

rdar://124913332

al45tair committed:
[Backtracing] Fix symbol lookup on Linux.

Also removed a spurious `print()` that got left in by accident,
and disabled the `CodableBacktrace.swift` test on Linux since we
need Foundation and we don't have it there.

rdar://124913332

al45tair commented:
@swift-ci Please test

al45tair commented:
This is still going to break because of the `-disable-implicit-backtracing-import` problem (since that's coming from other repos), but let's see how far we get otherwise.

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing][Tests] Improve debugability of CompactImageMap test.

If this test fails in CI, we could do with being able to see the
output from the test program in full.

rdar://124913332

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing] Fix compact image map decoding bug.

If we ended up with a `/` at the beginning of a string segment, we
were erroneously not adding to the expansion dictionary when we
should have been.

rdar://124913332

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing][Tests] Tweak swift-backtrace dependency.

We only want to add the `swift-backtrace` binary dependency when the
SDK we're looking at actually has `swift-backtrace` enabled.

rdar://124913332

al45tair commented:
@swift-ci Please test

weissi commented on stdlib/public/RuntimeModule/ImageMap+Darwin.swift:
```diff
@@ -0,0 +1,91 @@
+//
+//===----------------------------------------------------------------------===//
+
+#if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
```
Here and elsewhere: `visionOS` is missing. Could we make this `#if canImport(Darwin)` instead?

al45tair replied:
That's true, though the entire module only actually builds for macOS and Linux at present (which is something else we should probably think about). Previously we didn't use `#if canImport(Darwin)` because that actually didn't work — we didn't have the `Darwin` module by the time this code was being built, and I think from our end we'd really like to not use that pattern because there might be things that are not macOS/iOS etc that have a Darwin module.

al45tair replied:
(I'll add that I think it would be a good idea to make supporting this on iOS et al the subject of a separate PR. This one is already large enough.)

edymtt commented on stdlib/public/runtime/CMakeLists.txt:
```diff
@@ -144,7 +144,7 @@ foreach(sdk ${SWIFT_SDKS})
 endforeach()
 
-add_swift_target_library(swiftRuntime OBJECT_LIBRARY
+add_swift_target_library(swiftRuntimeCore OBJECT_LIBRARY
```
Would it be possible to add a comment to explain that we are resorting to rename this? Something akin to

```suggestion
# In modern CMake, we would be able to choose a different
# logical name for the Runtime module part of SE-0419, and change
# its target properties to reflect the module name.
# However, the current build system is architected in such a way
# that preclude the use of such properties, so we need to resort
# to rename the target for the Swift Runtime
add_swift_target_library(swiftRuntimeCore OBJECT_LIBRARY
```

al45tair committed:
[Backtracing] Add platform and architecture information.

It's useful to capture the platform and platform version with the image map.
Also, display both the platform and architecture information when generating
a crash log.

rdar://124913332

al45tair commented on docs/CompactImageMapFormat.md:
```diff
@@ -0,0 +1,226 @@
+
+If `e` is `1`, this opcode is followed by `code + 1` bytes that give
+a value `v` such that `v + 64` is the index into the prefix table for
+the prefix that should be appended to the string at this point.
```
The decoder _does_ support that, but the encoder presently does not — every path will be encoded as a sequence of expansions followed by a sequence of `str` instructions or a `framewk` instruction.

al45tair commented on stdlib/public/RuntimeModule/Address.swift:
```diff
@@ -0,0 +1,263 @@
+        return T(addr)
+      case let .sixtyFourBit(addr):
+        guard T.bitWidth >= 64 else { return nil }
+        return T(addr)
```
Requiring `UnsignedInteger` is tricky because this function exists in order to add an `init()` to `FixedWidthInteger` in an extension. But I think adding `truncatingIfNeeded:` should solve the problem in a reasonable manner; if someone is using signed addresses (and some processors do), it might make sense to convert to/from signed.

mikeash replied:
That sounds reasonable to me.

al45tair committed:
[Backtracing] Updates following review.

Added some explanatory text to the Compact Backtrace Format
documentation, and also to the `CMakeLists.txt` for the runtime.

Tweaked the conversions for `Backtrace.Address` to truncate, which
should result in reasonable behaviour for negative fixed-width
integers.

Use a constant for the ASCII values for `/` and `\` in the Compact
Image Format encoder/decoder.

Make the TSD key for `ElfImageCache` non-optional, and call
`fatalError()` if we don't get one.

rdar://124913332

al45tair commented:
@swift-ci Please test

edymtt reviewed (approved):
Looks good to me regarding the build system changes.
